### PR TITLE
Refactor type propagator and improve scope handling

### PR DESCRIPTION
**Pull request by kasperl**

### CODE REVIEW

**Review by floitsch — APPROVED** [review]
LGTM.

File: src/compiler/propagation/type_scope.cc [hunk 1]
```diff
@@ -0,0 +1,109 @@
...
+      type.fill(words_per_type_);
+    } else {
+      type.add(argument_type.id());
+    }
+  }
+}
+
+TypeScope::TypeScope(BlockTemplate* block, TypeScope* outer)
+    : words_per_type_(outer->words_per_type_)
+    , stacks_(static_cast<TypeStack**>(malloc((outer->level() + 2) * sizeof(TypeStack*))))
```
[the remark this answers is not in the document]
Comment (kasperl) [reply src/compiler/propagation/type_scope.cc:48]: Level 0 needs 1 slot.

File: src/compiler/propagation/type_propagator.cc [hunk 2]
```diff
@@ -857,10 +735,13 @@ static void process(MethodTemplate* method, uint8* bcp, TypeStack* stack, Workli
     Method target(program->bytecodes, program->dispatch_table[offset]);
     propagator->call_static(method, stack, bcp, target);
     if (stack->local(0).is_empty(program)) return;
-    if (stack->outer()) {
+    if (scope->level() > 0) {
```
[the remark this answers is not in the document]
Comment (kasperl) [reply src/compiler/propagation/type_propagator.cc:738]: I think that's right.

File: src/compiler/propagation/type_scope.cc [hunk 3]
```diff
@@ -0,0 +1,109 @@
...
+    , stacks_(static_cast<TypeStack**>(malloc(1 * sizeof(TypeStack*))))
+    , level_(0)
+    , outer_(null) {
+  int sp = method->method().arity() + Interpreter::FRAME_SIZE;
+  TypeStack* stack = new TypeStack(sp - 1, sp + method->method().max_height() + 1, words_per_type_);
+  stacks_[0] = stack;
+
+  for (unsigned i = 0; i < method->arity(); i++) {
+    TypeSet type = stack->get(i);
+    ConcreteType argument_type = method->argument(i);
```
[the remark this answers is not in the document]
Comment (kasperl) [reply src/compiler/propagation/type_scope.cc:35]: In one case the argument type is a ConcreteType. In the other, it is a TypeSet.

File: src/compiler/propagation/type_propagator.cc [hunk 4]
```diff
@@ -321,58 +321,54 @@ void TypePropagator::store_field(MethodTemplate* user, TypeStack* stack, int ind
...
-  TypeStack* outer = stack->outer();
-  int n = outer->level() - block.block()->level();
-  for (int i = 0; i < n; i++) outer = outer->outer();
-  TypeSet value = outer->local(index);
+  TypeSet value = scope->load_outer(block, index);
   stack->pop();
   stack->push(value);
   if (value.is_block()) return;
   // We keep track of the types we've seen for outer locals for
   // this particular access site. We use this to exclusively
```
[the remark this answers is not in the document]
Comment (kasperl) [reply src/compiler/propagation/type_propagator.cc]: Fixed.

**Pull request merged**